File: sdk/js/src/client.test.ts
import { describe, it, expect, vi } from 'vitest';
import { createClient, CuttlegateError } from './client.js';
import type { CuttlegateClient } from './client.js';

const validConfig = {
  baseUrl: 'https://cuttlegate.example.com',
  token: 'svc_abc',
  project: 'my-project',
  environment: 'production',
};

function mockFetchResponse(body: unknown, status = 200): typeof fetch {
  return vi.fn().mockResolvedValue({
    ok: status >= 200 && status < 300,
    status,
    json: () => Promise.resolve(body),
  });
}

const bulkResponse = {
  flags: [
    { key: 'dark-mode', enabled: true, value: null, value_key: 'true', reason: 'rule_match', type: 'bool' },
    { key: 'banner-text', enabled: true, value: 'holiday', value_key: 'holiday', reason: 'default', type: 'string' },
  ],
  evaluated_at: '2026-03-21T10:00:00Z',
};

describe('createClient', () => {
  it('returns a CuttlegateClient with valid config', () => {
    const client = createClient({ ...validConfig, fetch: mockFetchResponse({}) });
    expect(client).toBeDefined();
    expect(typeof client.evaluate).toBe('function');
    expect(typeof client.evaluateFlag).toBe('function');
  });

  it('throws synchronously when token is missing', () => {
    expect(() =>
      createClient({ ...validConfig, token: '' }),
    ).toThrow('token is required');
  });

  it('throws synchronously when baseUrl is missing', () => {
    expect(() =>
      createClient({ ...validConfig, baseUrl: '' }),
    ).toThrow('baseUrl is required');
  });

  it('throws synchronously when project is missing', () => {
    expect(() =>
      createClient({ ...validConfig, project: '' }),
    ).toThrow('project is required');
  });

  it('throws synchronously when environment is missing', () => {
    expect(() =>
      createClient({ ...validConfig, environment: '' }),
    ).toThrow('environment is required');
  });

  it('uses default timeout and native fetch when not provided', () => {
    const client = createClient(validConfig);
    expect(client).toBeDefined();
  });

  it('accepts optional timeout and custom fetch', () => {
    const customFetch = (() => {}) as unknown as typeof fetch;
    const client = createClient({
      ...validConfig,
      timeout: 10000,
      fetch: customFetch,
    });
    expect(client).toBeDefined();
  });

  it('is usable as a CuttlegateClient type', () => {
    function useClient(c: CuttlegateClient): boolean {
      return typeof c.evaluate === 'function' && typeof c.evaluateFlag === 'function';
    }
    const client = createClient({ ...validConfig, fetch: mockFetchResponse({}) });
    expect(useClient(client)).toBe(true);
  });

  it('does not expose the token on the returned client object', () => {
    const client = createClient({ ...validConfig, fetch: mockFetchResponse({}) });
    const serialised = JSON.stringify(client);
    expect(serialised).not.toContain('svc_abc');

    for (const value of Object.values(client)) {
      if (typeof value === 'string') {
        expect(value).not.toContain('svc_abc');
      }
    }
  });

  it('throws CuttlegateError with code field on validation failure', () => {
    try {
      createClient({ ...validConfig, baseUrl: '' });
      expect.fail('should have thrown');
    } catch (err) {
      expect(err).toBeInstanceOf(CuttlegateError);
      expect((err as CuttlegateError).code).toBe('invalid_config');
    }
  });

  it('error messages do not leak the token value', () => {
    const sensitiveToken = 'super_secret_token_12345';
    try {
      createClient({ baseUrl: '', token: sensitiveToken, project: 'p', environment: 'prod' });
      expect.fail('should have thrown');
    } catch (err) {
      const message = (err as Error).message;
      expect(message).not.toContain(sensitiveToken);
      expect(String(err)).not.toContain(sensitiveToken);
    }
  });
});

describe('evaluate', () => {
  it('returns all flags for a user context', async () => {
    const fetchFn = mockFetchResponse(bulkResponse);
    const client = createClient({ ...validConfig, fetch: fetchFn });

    const results = await client.evaluate({ user_id: 'u1', attributes: { plan: 'pro' } });

    expect(fetchFn).toHaveBeenCalledOnce();
    const [url, opts] = (fetchFn as ReturnType<typeof vi.fn>).mock.calls[0];
    expect(url).toBe('https://cuttlegate.example.com/api/v1/projects/my-project/environments/production/evaluate');
    expect(opts.method).toBe('POST');
    expect(opts.headers['Authorization']).toBe('Bearer svc_abc');
    expect(JSON.parse(opts.body)).toEqual({ context: { user_id: 'u1', attributes: { plan: 'pro' } } });

    expect(results).toHaveLength(2);
    expect(results[0]).toEqual({
      key: 'dark-mode',
      enabled: true,
      value: null,
      valueKey: 'true',
      reason: 'rule_match',
      evaluatedAt: '2026-03-21T10:00:00Z',
    });
    expect(results[1]).toEqual({
      key: 'banner-text',
      enabled: true,
      value: 'holiday',
      valueKey: 'holiday',
      reason: 'default',
      evaluatedAt: '2026-03-21T10:00:00Z',
    });
  });

  it('throws CuttlegateError with code="network_error" when fetch fails', async () => {
    const fetchFn = vi.fn().mockRejectedValue(new TypeError('Failed to fetch'));
    const client = createClient({ ...validConfig, fetch: fetchFn });

    await expect(client.evaluate({ user_id: 'u1', attributes: {} })).rejects.toSatisfy(
      (err: unknown) => err instanceof CuttlegateError && err.code === 'network_error' && err.message === 'Failed to fetch',
    );
  });

  it('throws CuttlegateError with code="timeout" when request times out', async () => {
    const fetchFn = vi.fn().mockImplementation((_url: string, opts: RequestInit) => {
      return new Promise((_resolve, reject) => {
        opts.signal?.addEventListener('abort', () => {
          reject(new DOMException('The operation was aborted.', 'AbortError'));
        });
      });
    });
    const client = createClient({ ...validConfig, timeout: 50, fetch: fetchFn });

    await expect(client.evaluate({ user_id: 'u1', attributes: {} })).rejects.toSatisfy(
      (err: unknown) => err instanceof CuttlegateError && err.code === 'timeout',
    );
  });

  it('throws CuttlegateError with code="unauthorized" on 401', async () => {
    const fetchFn = mockFetchResponse({}, 401);
    const client = createClient({ ...validConfig, fetch: fetchFn });

    await expect(client.evaluate({ user_id: 'u1', attributes: {} })).rejects.toSatisfy(
      (err: unknown) => err instanceof CuttlegateError && err.code === 'unauthorized',
    );
  });

  it('throws CuttlegateError with code="forbidden" on 403', async () => {
    const fetchFn = mockFetchResponse({}, 403);
    const client = createClient({ ...validConfig, fetch: fetchFn });

    await expect(client.evaluate({ user_id: 'u1', attributes: {} })).rejects.toSatisfy(
      (err: unknown) => err instanceof CuttlegateError && err.code === 'forbidden',
    );
  });

  it('throws CuttlegateError with code="invalid_response" on malformed JSON shape', async () => {
    const fetchFn = mockFetchResponse({ unexpected: 'shape' });
    const client = createClient({ ...validConfig, fetch: fetchFn });

    await expect(client.evaluate({ user_id: 'u1', attributes: {} })).rejects.toSatisfy(
      (err: unknown) =>
        err instanceof CuttlegateError &&
        err.code === 'invalid_response' &&
        err.message === 'Server returned an unexpected response shape',
    );
  });
});

describe('evaluateFlag', () => {
  it('returns a single flag result from bulk response', async () => {
    const client = createClient({ ...validConfig, fetch: mockFetchResponse(bulkResponse) });

    const result = await client.evaluateFlag('dark-mode', { user_id: 'u1', attributes: {} });
    // @happy: valueKey is "true" for bool flag; value remains null (v1 compat)
    expect(result).toEqual({ enabled: true, value: null, valueKey: 'true', reason: 'rule_match' });
  });

  it('returns not_found for unknown key without throwing', async () => {
    const client = createClient({ ...validConfig, fetch: mockFetchResponse(bulkResponse) });

    const result = await client.evaluateFlag('beta-feature', { user_id: 'u1', attributes: {} });
    expect(result).toEqual({ enabled: false, value: null, valueKey: '', reason: 'not_found' });
  });

  it('makes only one HTTP call (wraps evaluate)', async () => {
    const fetchFn = mockFetchResponse(bulkResponse);
    const client = createClient({ ...validConfig, fetch: fetchFn });

    await client.evaluateFlag('dark-mode', { user_id: 'u1', attributes: {} });
    expect(fetchFn).toHaveBeenCalledOnce();
  });
});

File: sdk/go/client_test.go
package cuttlegate_test

import (
	"context"
	"encoding/json"
	"errors"
	"net/http"
	"net/http/httptest"
	"testing"

	cuttlegate "github.com/karo/cuttlegate/sdk/go"
)

// bulkResponse is a valid evaluate API response for use in tests.
var bulkResponse = map[string]any{
	"flags": []map[string]any{
		{"key": "dark-mode", "enabled": true, "value": nil, "value_key": "true", "reason": "rule_match", "type": "bool"},
		{"key": "banner-text", "enabled": true, "value": "holiday", "value_key": "holiday", "reason": "default", "type": "string"},
	},
	"evaluated_at": "2026-03-21T10:00:00Z",
}

func serveBulk(t *testing.T, body any, status int) *httptest.Server {
	t.Helper()
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		w.Header().Set("Content-Type", "application/json")
		w.WriteHeader(status)
		if err := json.NewEncoder(w).Encode(body); err != nil {
			t.Errorf("test server encode: %v", err)
		}
	}))
}

func validConfig(baseURL string) cuttlegate.Config {
	return cuttlegate.Config{
		BaseURL:      baseURL,
		ServiceToken: "svc_test",
		Project:      "my-project",
		Environment:  "production",
	}
}

// --- @happy: NewClient validation ---

func TestNewClient_ValidConfig(t *testing.T) {
	// @happy
	srv := serveBulk(t, bulkResponse, 200)
	defer srv.Close()

	c, err := cuttlegate.NewClient(validConfig(srv.URL))
	if err != nil {
		t.Fatalf("expected no error, got %v", err)
	}
	if c == nil {
		t.Fatal("expected non-nil client")
	}
}

func TestNewClient_MissingBaseURL(t *testing.T) {
	// @error-path
	_, err := cuttlegate.NewClient(cuttlegate.Config{
		ServiceToken: "tok",
		Project:      "proj",
		Environment:  "prod",
	})
	if err == nil {
		t.Fatal("expected error for missing BaseURL")
	}
}

func TestNewClient_MissingServiceToken(t *testing.T) {
	// @error-path
	_, err := cuttlegate.NewClient(cuttlegate.Config{
		BaseURL:     "http://localhost",
		Project:     "proj",
		Environment: "prod",
	})
	if err == nil {
		t.Fatal("expected error for missing ServiceToken")
	}
}

func TestNewClient_MissingProject(t *testing.T) {
	// @error-path
	_, err := cuttlegate.NewClient(cuttlegate.Config{
		BaseURL:      "http://localhost",
		ServiceToken: "tok",
		Environment:  "prod",
	})
	if err == nil {
		t.Fatal("expected error for missing Project")
	}
}

func TestNewClient_MissingEnvironment(t *testing.T) {
	// @error-path
	_, err := cuttlegate.NewClient(cuttlegate.Config{
		BaseURL:      "http://localhost",
		ServiceToken: "tok",
		Project:      "proj",
	})
	if err == nil {
		t.Fatal("expected error for missing Environment")
	}
}

func TestNewClient_NoNetworkCallsAtInit(t *testing.T) {
	// @happy — validation must be local only
	called := false
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		called = true
		w.WriteHeader(200)
	}))
	defer srv.Close()

	_, err := cuttlegate.NewClient(validConfig(srv.URL))
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if called {
		t.Fatal("NewClient must not make network calls")
	}
}

// --- @happy: Evaluate ---

func TestEvaluate_ReturnsBulkResults(t *testing.T) {
	// @happy
	srv := serveBulk(t, bulkResponse, 200)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	results, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1", Attributes: map[string]any{"plan": "pro"}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(results) != 2 {
		t.Fatalf("expected 2 results, got %d", len(results))
	}
	if results[0].Key != "dark-mode" || !results[0].Enabled || results[0].Reason != "rule_match" {
		t.Errorf("unexpected result[0]: %+v", results[0])
	}
	// @happy: value_key is "true" for bool flag, Value is "" (nil in wire format)
	if results[0].ValueKey != "true" {
		t.Errorf("expected ValueKey=true for bool flag, got %q", results[0].ValueKey)
	}
	if results[1].Key != "banner-text" || results[1].Value != "holiday" {
		t.Errorf("unexpected result[1]: %+v", results[1])
	}
	// @happy: value_key equals value for string flag
	if results[1].ValueKey != "holiday" {
		t.Errorf("expected ValueKey=holiday for string flag, got %q", results[1].ValueKey)
	}
}

func TestEvaluate_SetsAuthorizationHeader(t *testing.T) {
	// @happy
	var gotAuth string
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		gotAuth = r.Header.Get("Authorization")
		w.Header().Set("Content-Type", "application/json")
		json.NewEncoder(w).Encode(bulkResponse)
	}))
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"})

	if gotAuth != "Bearer svc_test" {
		t.Errorf("expected Bearer svc_test, got %q", gotAuth)
	}
}

// --- @auth-bypass: 401/403 ---

func TestEvaluate_Returns_AuthError_On401(t *testing.T) {
	// @auth-bypass
	srv := serveBulk(t, map[string]string{"error": "unauthorized"}, 401)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	_, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"})

	var authErr *cuttlegate.AuthError
	if !errors.As(err, &authErr) {
		t.Fatalf("expected *AuthError, got %T: %v", err, err)
	}
	if authErr.StatusCode != 401 {
		t.Errorf("expected StatusCode 401, got %d", authErr.StatusCode)
	}
}

func TestEvaluate_Returns_AuthError_On403(t *testing.T) {
	// @auth-bypass
	srv := serveBulk(t, map[string]string{"error": "forbidden"}, 403)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	_, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"})

	var authErr *cuttlegate.AuthError
	if !errors.As(err, &authErr) {
		t.Fatalf("expected *AuthError, got %T: %v", err, err)
	}
	if authErr.StatusCode != 403 {
		t.Errorf("expected StatusCode 403, got %d", authErr.StatusCode)
	}
}

// --- @error-path: ServerError ---

func TestEvaluate_Returns_ServerError_On500(t *testing.T) {
	// @error-path
	srv := serveBulk(t, map[string]string{"error": "internal"}, 500)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	_, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"})

	var srvErr *cuttlegate.ServerError
	if !errors.As(err, &srvErr) {
		t.Fatalf("expected *ServerError, got %T: %v", err, err)
	}
	if srvErr.StatusCode != 500 {
		t.Errorf("expected StatusCode 500, got %d", srvErr.StatusCode)
	}
}

// --- @error-path: NotFoundError ---

func TestEvaluate_Returns_NotFoundError_On404(t *testing.T) {
	// @error-path
	srv := serveBulk(t, map[string]string{"error": "not_found"}, 404)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	_, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"})

	var nfErr *cuttlegate.NotFoundError
	if !errors.As(err, &nfErr) {
		t.Fatalf("expected *NotFoundError, got %T: %v", err, err)
	}
	if nfErr.Resource != "project" {
		t.Errorf("expected Resource=project, got %q", nfErr.Resource)
	}
}

// --- @happy: EvaluateFlag ---

func TestEvaluateFlag_ReturnsSingleFlag(t *testing.T) {
	// @happy
	srv := serveBulk(t, bulkResponse, 200)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	result, err := c.EvaluateFlag(context.Background(), "dark-mode", cuttlegate.EvalContext{UserID: "u1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if !result.Enabled || result.Reason != "rule_match" {
		t.Errorf("unexpected result: %+v", result)
	}
	// @happy: value_key is present on FlagResult
	if result.ValueKey != "true" {
		t.Errorf("expected ValueKey=true for bool flag, got %q", result.ValueKey)
	}
}

func TestEvaluateFlag_ReturnsNotFoundReasonForMissingKey(t *testing.T) {
	// @edge
	srv := serveBulk(t, bulkResponse, 200)
	defer srv.Close()

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	result, err := c.EvaluateFlag(context.Background(), "nonexistent-flag", cuttlegate.EvalContext{UserID: "u1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if result.Enabled {
		t.Error("expected Enabled=false for missing key")
	}
	if result.Reason != "not_found" {
		t.Errorf("expected Reason=not_found, got %q", result.Reason)
	}
}

// --- @edge: context cancellation ---

func TestEvaluate_RespectsContextCancellation(t *testing.T) {
	// @edge
	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		// Block until the client cancels.
		<-r.Context().Done()
		w.WriteHeader(200)
	}))
	defer srv.Close()

	ctx, cancel := context.WithCancel(context.Background())
	cancel() // cancel immediately

	c, _ := cuttlegate.NewClient(validConfig(srv.URL))
	_, err := c.Evaluate(ctx, cuttlegate.EvalContext{UserID: "u1"})
	if err == nil {
		t.Fatal("expected error for cancelled context")
	}
}

// --- @happy: custom HTTPClient ---

func TestNewClient_AcceptsCustomHTTPClient(t *testing.T) {
	// @happy
	srv := serveBulk(t, bulkResponse, 200)
	defer srv.Close()

	cfg := validConfig(srv.URL)
	cfg.HTTPClient = &http.Client{}
	c, err := cuttlegate.NewClient(cfg)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if _, err := c.Evaluate(context.Background(), cuttlegate.EvalContext{UserID: "u1"}); err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}
